Adds Dropaligner, Dropaccelerator and FixRegion plugins

JoshuaMarx opened the pull request:
Dropaligner: This plugin is used to fix a liquid droplet within a vapor phase in its position.

It is designed in analogy with the COMaligner. The user specifies the desired center of mass and the radius within which the realignment takes place.

No known issues

Dropaccelerator:

This plugin is used to accelerate a spherical droplet. 
The user specifies the center of mass and radius of the droplet. 
All molecules within this sphere are then marked with the bool variable DROP = true in FullMolecule.
To this selection of molecules a velocity increment is added every time step for a specified amount of steps to speed up the drop to a desired velocity.
The drop velocity is tracked for some time and increased if the desired velocity wasn't reached within the specified time frame.

Issues: After acceleration, the velocity of the droplet is higher than the desired velocity. I observed that the number of particles marked as DROP == true is not constant, which should not be possible if the plugin was operating properly. Is this approach with marking particles using bool variables viable at all? I am also not certain whether beforeForces is the proper function to use here. At which point during a time step does it make the most sense to accelerate particles?

FixRegion:

FixRegion: 

Function: This plugin is supposed to keep particles within a specified volume from moving during simulation. The particles still interact with non-fixed particles via forces.
Specifically, this plugin is supposed to fix an atomistic wall, by immobilizing the bottom layers of the wall.
These bottom layers of the wall hold the rest of the wall in place via force interactions.

Implementation: Every time step the force and velocity of the particles within the specified volume is set to zero.

Problems: In test simulations of a sessile droplet on the wall, the droplet exhibits an unusual pressure profile.
I am not sure what causes this. I previously tried a similar approach as in the Dropaccelerator plugin, wherein I marked the particles of interest with a bool.
I then excluded these particles from the leapfrog integration. This approach also caused problems as the number of marked particles did not remain constant.
The current implementation however causes the aforementioned problems with the pressure. 
Is this way of realizing a fixed region sensible? At which point during a time step would it make the most sense to set velocities and forces to zero (if at all)?

SteffenSeckler commented on src/plugins/Dropaccelerator.cpp:
```diff
@@ -0,0 +1,190 @@
+            domainDecomp->collCommFinalize();
+
+            // CALCULATE AVERAGE SPEED
+            _VelocNOW = -(_VelocNOW / Particles);
```
why `-`?

SteffenSeckler commented on src/plugins/Dropaccelerator.cpp:
```diff
@@ -0,0 +1,190 @@
+
+        if (simstep >= (_StartSimstep + _Steps) && simstep < (_StartSimstep + 2 * _Steps)) {
+            _VelocNOW = 0;
+            int Particles = 0;
```
please rename to `numParticlesInDroplet` or so.

SteffenSeckler commented on src/plugins/Dropaccelerator.cpp:
```diff
@@ -0,0 +1,190 @@
+    DomainDecompBase* domainDecomp,
+    unsigned long simstep) {
+
+    double corrVeloc = -(_Veloc/_Steps);
```
why `-`? can't you add this to the input file? that way everything would be cleaner.

JoshuaMarx replied:
Yes I guess we'll add this to the input. I will be discussing all of these plugins in detail with Fabio later today, since there are obviously still some general problems with functionality,

JoshuaMarx commented:
@SteffenSeckler Die global communication Funktionen gibt es für bools nicht oder? Vielleicht _particleIsInDroplet besser als integer deklarieren?

SteffenSeckler commented:
you're free to add support for `bool`s or use MPI functions itself. Changing the data type, because a function is not yet included, does not make too much sense...

SteffenSeckler commented:
I've implemented it for you, as it also required a change to std::vector<char>.
std::vector<bool> might use a specialized variant of std::vector which cannot be easily transferred via mpi.

JoshuaMarx commented:
> 
> 
> I've implemented it for you, as it also required a change to std::vector.
> std::vector might use a specialized variant of std::vector which cannot be easily transferred via mpi.

Thank you! I implemented collCommGetBool and collCommAppendBool meanwhile although I am not sure it makes much sense since there is no LOR yet for collective communication, only SUM and I am not sure how to implement it. Should I upload it anyway for possible future applications or just dispense with it?

SteffenSeckler commented:
dispense :)
we normally encourage the use of vectors and the collcomm is anyways not meant for large arrays, as everything is copied twice.

JoshuaMarx commented on src/plugins/FixRegion.cpp:
```diff
@@ -46,16 +46,16 @@ void FixRegion::init(ParticleContainer* particleContainer, DomainDecompBase* dom
-				tm->setF(i, 0.0);
-			}
-			_molCount++;
+	for (auto tm = particleContainer->regionIterator(min.data(), max.data(), ParticleIterator::ONLY_INNER_AND_BOUNDARY);
```
OK, so I did a bunch of tests and this part here didn't work. What happens is that the regionIterator basically doesn't do anything, i.e. no particles are examined and then the simulation crashes when attempting to do the global communication step for the variable _molCount. We then tried to rework this part, although I am not sure I defined the borders of the region correctly since it still does not work: 

```
double regionLowCorner[3], regionHighCorner[3];
	
	if(particleContainer->getBoundingBoxMin(0)>=_xMin && particleContainer->getBoundingBoxMin(1)>=_yMin && particleContainer->getBoundingBoxMin(2)>=_zMin &&
		particleContainer->getBoundingBoxMax(0)<=_xMax && particleContainer->getBoundingBoxMax(1)<=_yMax && particleContainer->getBoundingBoxMax(2)<=_zMax) {
			for(unsigned d = 0; d < 3; d++){
				regionLowCorner[d] = particleContainer->getBoundingBoxMin(d);
				regionHighCorner[d] = particleContainer->getBoundingBoxMax(d);
			}
		} 
#if defined (_OPENMP)
#pragma omp parallel shared(regionLowCorner, regionHighCorner)
#endif
	
	// ITERATE OVER PARTICLES
	for (auto temporaryMolecule = particleContainer->regionIterator(regionLowCorner, regionHighCorner, ParticleIterator::ALL_CELLS);
		 temporaryMolecule.isValid(); ++temporaryMolecule) {
		for (unsigned i = 0; i < 3; i++) {
			temporaryMolecule->setv(i, 0.0);
			temporaryMolecule->setF(i, 0.0);
		}
		global_log->info() << "A" << std::endl;
		temporaryMolecule->clearFM();
		_molCount++;
	}
```

Could you have a look @SteffenSeckler ?

SteffenSeckler replied:
see above. you also don't want to use ParticleIterator::ALL_CELLS.
